File: python/basics/binary_search/69_square_root.py
'''
https://leetcode.com/problems/sqrtx/description/
Approach:
    use binary search.. iterate from 1..n.  
    binary search ... test if its square root..
'''

def square_root(num):
    if num < 2:
        return num

    l = 2
    r = num // 2
    while l <= r:
        m = l + (r-l)//2
        sq = m * m
        if sq == num:
            return m
        elif sq > num:  #mistake there was a typo here
            r = m -1
        else:
            l = m + 1
    return r

x = 8
result = square_root(x)
print("square root of : " + str(x) + "      " + str(result))


x = 2
result = square_root(x)
print("square root of : " + str(x) + "      " + str(result))


x = 0
result = square_root(x)
print("square root of : " + str(x) + "      " + str(result))


x = 4
result = square_root(x)
print("square root of : " + str(x) + "      " + str(result))

x = 225
result = square_root(x)
print("square root of : " + str(x) + "      " + str(result))



    

File: problem_solving/binary_search/cpp/69_square_root.cpp
/*
https://leetcode.com/problems/sqrtx/description/

use long long
*/

int my_sqrt(int x) {
    if(x < 2) {
        return x;
    }
    int l = 1;
    int r = x / 2;
    while(l <= r) {
        long long m = l + (r-l)/2;
        long long sq = m*m;
        if(sq == x) {
            return m;
        } else if(sq < x) {
            l = m + 1;
        } else {
            r = m-1;            
        }
    }
    return r;
}


int main() {
    return 0;
}